enhancements/machine-config/mco-state-reporting.md

```diff
@@ -0,0 +1,499 @@
+---
+title: mco-state-reporting
+authors:
+  - "@cdoern"
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - "@sinnykumari" # MCO
+  - "@yuqi-zhang" # MCO
+approvers:
+  - "@sinnykumari"
+  - "@yuqi-zhang"
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - "@JoelSpeed"
+creation-date: 2023-10-05
+last-updated: 2023-10-12
+tracking-link:
+  - https://issues.redhat.com/browse/MCO-452
+see-also:
+replaces:
+superseded-by:
+---
+
+# Track MCO State Related to Upgrades and Everyday Operations.
+
+## Summary
+
+This enhancement describes how Nodes and their upgrade processes should be aggregated by the MCO into a user facing object. Alongside this, MCO component progression should be aggregated into the machineconfiguration operator/v1 object. The goal here is to allow customers and the MCO team to decipher the our processes in a more verbose way, speeding up the debugging process and allowing for better customer engagement.
+
+## Motivation
+
+The MCO manages node upgrades but since we do not own the object or store much of their data in other ways, much of what occurs during an upgrade
+is simply a black box operation that we currently report as "Updating" or "Updated". Users can debug into a specific node or look into the node spec for some of this information
+but most of it simply lives in the MCO code rather than in data structures. We want to put these abstract "phases" of node operations as triggered by the MCO into a concrete data structure. 
+
+The MCO also has "day to day" operations on its MachineConfigController, MachineConfigDaemon, and MachineConfigOperator pods. Tracking these daily progression statuses is just as important as the upgrade status. However, this information does not belong in the upgrade-progression datatype. Currently, only the MCO team has the skillset to dive into the code and figure out what "phase" of the MCO causes a non-upgrade related error. The goal of this work is to make that information accessible and understandable to an everyday audience.
+
+### User Stories
+
+* As a <role>, I want to <take some action> so that I can <accomplish a
+goal>.
+
+### Goals
+
+* Make a MachineConfigState type that succinctly holds the upgrade data.
```

Comment:
Would a MC / MCP logic to maintain observedGeneration fit the enhancement theme too? The idea is that higher level operators (or humans) might want to know whether MCO noticed a change to a object and updated a status reflecting it. I noticed it is mentioned below, but not here in the goals.

For example:

1. The user updates a MachineConfig (this changes the generation counter)
2. Here MC / MCP are still showing Updating=false, but it is not clear whether that is stale status or MCO was really that quick (but with observedGeneration in place the MC would have generation=X and observedGeneration=X-1)
3. (enhanced) MCO updates the observedGeneration of the MC / MCP to match the generation AND the statuses/conditions (Updating=true) to reflect that something is happening

This would make observability and composability much simpler for tools like ACM and others.

Reply:
Yes, this is something that is going live in both the machineconfiguration operator level object as well as the upgrade progression tracking. the machineconfiguration object would detect the initial flurry of changes in the operator pod followed by the upgrade progression being kicked off by a change in machineconfig.